static/app/utils/profiling/renderers/flamegraphRenderer.tsx

```diff
@@ -0,0 +1,675 @@
+        type,
+        normalize,
+        stride,
+        offset
+      );
+      // Point to attribute location
+      this.gl.enableVertexAttribArray(aBoundsAttributeLocation);
+    }
+
+    // Use shader program
+    this.gl.useProgram(this.program);
+  }
+
+  setConfigView(configView: Rect): Rect {
+    this.configView = computeClampedConfigView(
+      configView,
+      {
+        min: this.flamegraph.profile.minFrameDuration,
+        max: this.configSpace.width,
+      },
+      {
+        min: 0,
+        max: this.configSpace.height,
+      },
+      !!this.flamegraph.inverted
+    );
+    return this.configView;
+  }
+
+  transformConfigView(transformation: mat3): Rect {
+    const newConfigViewSpace = this.configView.transformRect(transformation);
+
+    this.configView = computeClampedConfigView(
+      newConfigViewSpace,
+      {
+        min: this.flamegraph.profile.minFrameDuration,
+        max: this.configSpace.width,
+      },
+      {
+        min: 0,
+        max: this.configSpace.height,
+      },
+      !!this.flamegraph.inverted
+    );
+
+    return this.configView;
+  }
+
+  getColorForFrame(frame: Frame): number[] {
+    return (
+      this.colorMap.get(frame.name + (frame.file ? frame.file : '')) ??
+      this.theme.COLORS.FRAME_FALLBACK_COLOR
+    );
+  }
+
+  getConfigSpaceCursor(
+    logicalSpaceCursor: vec2,
+    configToPhysicalSpace: mat3 = this.configToPhysicalSpace
+  ): vec2 {
+    const physicalSpaceCursor = vec2.transformMat3(
+      vec2.create(),
+      logicalSpaceCursor,
+      this.logicalToPhysicalSpace
+    );
+
+    const physicalToConfig = mat3.invert(mat3.create(), configToPhysicalSpace);
+
+    return vec2.transformMat3(vec2.create(), physicalSpaceCursor, physicalToConfig);
+  }
+
+  getHoveredNode(configSpaceCursor: vec2): FlamegraphFrame | null {
+    let hoveredNode: FlamegraphFrame | null = null;
+
+    const findHoveredNode = (frame: FlamegraphFrame, depth: number) => {
+      const frameRect = new Rect(
+        frame.start,
+        this.flamegraph.inverted ? this.configSpace.height - frame.depth : frame.depth,
+        frame.end - frame.start,
+        1
+      );
+
+      // We treat entire flamegraph as a segment tree, this allows us to query in O(log n) time by
+      // only looking at the nodes that are relevant to the current cursor position. We discard any values
+      // on x axis that do not overlap the cursor, and descend until we find a node that overlaps at cursor y position
+      if (!frameRect.containsX(configSpaceCursor)) {
+        return;
+      }
+
+      // If our frame depth overlaps cursor y position, we have found our node
+      if (frameRect.containsY(configSpaceCursor)) {
+        hoveredNode = frame;
+        return;
+      }
+
+      // Descend into the rest of the children
+      for (const child of frame.children) {
+        findHoveredNode(child, depth + 1);
+      }
+    };
+
+    for (const r of this.roots) {
+      findHoveredNode(r, 0);
+    }
```

Comment:
Yeah, in reality tho, they won't go more than depth=1, because vector.x will not overlap with the root rect - I will early return if hoveredNode is found tho, let's keep it as explicit as possible